Functionalised join_intersect.

OceanNuclear opened the pull request:
## Linked Issues

Closes #3926 

## Description
Changed the signature of join_intersect.

## Interface Changes

join_intersect has been changed. But looks like no end-user will see/use this function directly.

## Checklist

I confirm that I have completed the following checks:

- [x] Tests run locally and pass `pytest tests --reactor`
- [x] Code quality checks run locally and pass `pre-commit run --from-ref develop --to-ref HEAD`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

je-cook commented on bluemira/equilibria/flux_surfaces.py:
```diff
@@ -136,8 +135,14 @@ def plot(self, ax: plt.Axes | None = None, **kwargs):
+            An instance of FluxSurface, with exactly the same class and coordinates,
+            but sharing a different underlying copy of the data (Coordinates).
+        """
+        return self.__class__(self.coords.copy())
```
```suggestion
        return type(self)(self.coords.copy())
```

je-cook commented on bluemira/equilibria/flux_surfaces.py:
```diff
@@ -477,7 +481,8 @@ def clip(self, first_wall: Coordinates):
         # Because we oriented the Coordinates the "right" way, the first intersection
         # is at the smallest argument
-        self.coords = Coordinates(self.coords[:, : min(args) + 1])
+        shortened_coords = coords_union_inters[:, : min(args) + 1]
+        self.coords = Coordinates(shortened_coords)
```
```suggestion
        self.coords = Coordinates(coords_union_inters[:, : min(args) + 1])
```

je-cook commented on bluemira/geometry/coordinates.py:
```diff
@@ -1779,6 +1779,17 @@ def __iter__(self):
+        An instance of Coordinates, with exactly the same class and coordinates,
+        but sharing a different underlying copy of the data array.
+        """
+        return self.__class__(self.as_dict())
```
This is a bit dangerous because the underlying arrays are not copied therefore the whole stack of copied coords is mutable. Also 
```suggestion
        return type(self)(self.as_dict())
```

je-cook reviewed (changes requested):
One pretty dangerous point that needs to be covered eg copying and changing the copy might change the original. We need to disconnect the numpy array pointers

OceanNuclear commented:
Not sure the reason for type(self) over self.__class__ (I suppose it's something about unifying the style?), but I've implemented that regardless. I thought as_dict would've handled the copying, but I guess not, fixed that as well now :)

je-cook commented:
> Not sure the reason for type(self) over self.**class** (I suppose it's something about unifying the style?), but I've implemented that regardless.

Its semi stylistic, on the whole we should avoid using dunder methods in code or at least minimise their usage. Using type directly makes sure that any odd edge cases are not our responsibility.

OceanNuclear commented:
Have checked that all copying functions make copies of the numpy matrix, rather than a pointer to the matrix itself/ a slice of the matrix etc.
